### PR TITLE
Feature/enhance engagement notification system

### PR DESCRIPTION
**Pull request by BotondAKovacs**
## Summary:

- Make engagement notification monitoring independent of around you screen
- It is now launched when the user clicks on sign in, stopped when the user logs out or closes the app (in background it should stay running)

## Futur Improvement:
- right now users can be spammed if they enter a zone with lots of people since the amount of notifications a client receives isn't limited
- also restarting the app can erase the map that avoid receiving the same notification until a cooldown period, this map could be stored and retrieved from the datastore

### CODE REVIEW

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 1]
```diff
@@ -78,19 +90,51 @@ class EngagementNotificationManager(
    * and processes them if the user is discoverable.
    */
   private fun checkNearbyUsersForCommonTags() {
-    profilesViewModel.getSelfProfile {
-      val selfProfile = profilesViewModel.selfProfile.value ?: return@getSelfProfile
-      val selfLocation = selfProfile.location ?: return@getSelfProfile
+    Log.e(TAG, "Checking for nearby users")
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:93]: be aware that you have a lot of logs in this file that do not indicate errors and that are logged very often, and might remove them for the final implementation

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 2]
```diff
@@ -78,19 +90,51 @@ class EngagementNotificationManager(
...
+    scope.launch {
+      try {
+        // Check if discoverable
+        val isDiscoverable = appDataStore.isDiscoverable.first()
+        Log.e(TAG, "User discoverable status: $isDiscoverable")
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:98]: log

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 3]
```diff
@@ -43,29 +43,41 @@ class EngagementNotificationManager(
...
+    private val tagsViewModel: TagsViewModel
 ) {
   private var notificationJob: Job? = null
-  private val scope = CoroutineScope(Dispatchers.Main)
+  private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
   private val lastNotificationTimes = mutableMapOf<String, Long>()
 
   /** Start monitoring for nearby users with common tags */
   fun startMonitoring() {
+    Log.e(TAG, "Starting engagement monitoring")
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:55]: log

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 4]
```diff
@@ -43,29 +43,41 @@ class EngagementNotificationManager(
...
     notificationJob =
         scope.launch {
-          // while (true) {
-          checkNearbyUsersForCommonTags()
-          // delay(CHECK_INTERVAL)
-          // }
+          Log.e(TAG, "Starting periodic checks")
+          try {
+            while (true) {
+              Log.e(TAG, "Running periodic check")
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:63]: log

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 5]
```diff
@@ -43,29 +43,41 @@ class EngagementNotificationManager(
...
+              checkNearbyUsersForCommonTags()
+              Log.e(TAG, "Waiting for next check interval (${CHECK_INTERVAL}ms)")
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:65]: log

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 6]
```diff
@@ -78,19 +90,51 @@ class EngagementNotificationManager(
...
+          profilesViewModel.getFilteredProfilesInRadius(
+              center = selfLocation,
+              radiusInMeters = filterViewModel.selectedRadius.value,
+              genders = filterViewModel.selectedGenders.value,
+              ageRange = filterViewModel.ageRange.value,
+              tags = tagsViewModel.filteredTags.value)
+
+          // Wait a bit for profiles to be fully loaded, then process them once
+          scope.launch {
+            delay(200) // Wait for profiles to stabilize
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:130]: nitpick: magic number

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 7]
```diff
@@ -78,19 +90,51 @@ class EngagementNotificationManager(
...
+            val nearbyProfiles = profilesViewModel.filteredProfiles.value
+            Log.e(TAG, "Found ${nearbyProfiles.size} nearby profiles")
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:132]: nitpick: log

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 8]
```diff
@@ -106,28 +150,41 @@ class EngagementNotificationManager(
    * the nearby user using the first common tag.
    */
   private fun processNearbyProfiles(selfProfile: Profile, nearbyProfiles: List<Profile>) {
+    Log.e(TAG, "Processing ${nearbyProfiles.size} nearby profiles")
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:153]: nitpick: log

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 9]
```diff
@@ -78,19 +90,51 @@ class EngagementNotificationManager(
...
+          if (selfProfile == null) {
+            Log.e(TAG, "Self profile is null")
+            return@getSelfProfile
+          }
+          val selfLocation = selfProfile.location
+          if (selfLocation == null) {
+            Log.e(TAG, "Self location is null")
+            return@getSelfProfile
+          }
+          Log.e(
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:116]: nitpick: log

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 10]
```diff
@@ -106,28 +150,41 @@ class EngagementNotificationManager(
...
     for (nearbyProfile in newListMinusSelf) {
-      // Skip if we've recently notified this user
-      val lastTime = lastNotificationTimes[nearbyProfile.uid] ?: 0L
-      if (System.currentTimeMillis() - lastTime < NOTIFICATION_COOLDOWN) continue
+      try {
+        // Check cooldown
+        val lastTime = lastNotificationTimes[nearbyProfile.uid] ?: 0L
+        val timeSinceLastNotification = System.currentTimeMillis() - lastTime
+        if (timeSinceLastNotification < NOTIFICATION_COOLDOWN) {
+          Log.e(
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:164]: nitpick: log

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 11]
```diff
@@ -106,28 +150,41 @@ class EngagementNotificationManager(
...
-        } catch (e: Exception) {
-          Log.e("EngagementNotification", "Failed to send notification", e)
+        if (commonTags.isNotEmpty()) {
+          val commonTag = commonTags.first()
+          Log.e(TAG, "Sending notification for common tag: $commonTag")
+          try {
+            meetingRequestViewModel.engagementNotification(
+                targetToken = nearbyProfile.fcmToken ?: "null", tag = commonTag)
+            lastNotificationTimes[nearbyProfile.uid] = System.currentTimeMillis()
+            Log.e(TAG, "Successfully sent notification to ${nearbyProfile.uid}")
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:181]: nitpick: log

File: app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt [hunk 12]
```diff
@@ -106,28 +150,41 @@ class EngagementNotificationManager(
...
+          } catch (e: Exception) {
+            Log.e(TAG, "Failed to send notification to ${nearbyProfile.uid}: ${e.message}", e)
```
Comment (louiscrc) [inline app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt:183]: nitpick: log

**Comment by louiscrc** [comment]
Very nice PR @BotondAKovacs 

Just a few things you can correct:  
- logs: either remove them, or if you want to maintain them in your code, be aware to use the right ones (you regularly use `Log.e` which is for errors instead of `Log.i`/ `Log.v`
- the notifications are still very long to appear on the phone despite being sent early on firebase ; as you are concerned by the problem, take a closer look at this problem to see if you can fix it in your implementation

Thank you !

**Review by louiscrc — CHANGES_REQUESTED** [review]
Very nice PR @BotondAKovacs

Just a few things you can correct:

logs: either remove them, or if you want to maintain them in your code, be aware to use the right ones (you regularly use `Log.e` which is for errors instead of `Log.i/ Log.v`
the notifications are still very long to appear on the phone despite being sent early on firebase ; as you are concerned by the problem, take a closer look at this problem to see if you can fix it in your implementation

Thank you !

**Comment by BotondAKovacs** [comment]
> logs: either remove them, or if you want to maintain them in your code, be aware to use the right ones (you regularly use `Log.e` which is for errors instead of `Log.i/ Log.v` 

I think some of them can be useful, I'll try to keep the most useful ones and you're right that I shouldn't use Log.e

**Comment by BotondAKovacs** [comment]
> the notifications are still very long to appear on the phone despite being sent early on firebase ; as you are concerned by the problem, take a closer look at this problem to see if you can fix it in your implementation

It's only on one of the emulators for you too right? I have no idea, maybe it is a emulator overload problem, in the logs they are sent at the exact same time. I don't know what to do with it

**Comment by louiscrc** [comment]
Nice PR Botond ! 
Thank you for implementing my remarks about logging, and your coverage is good !
LGTM 👍

**Review by louiscrc — APPROVED** [review]

**Pull request merged**